console/collections/src/merkle_tree/tests/update.rs

```diff
@@ -0,0 +1,520 @@
+            let proof = merkle_tree.prove(leaf_index, leaf)?;
+            // Verify the Merkle proof succeeds.
+            assert!(proof.verify(leaf_hasher, path_hasher, merkle_tree.root(), leaf));
+            // Verify the Merkle proof **fails** on an invalid root.
+            assert!(!proof.verify(leaf_hasher, path_hasher, &PH::Hash::zero(), leaf));
+            assert!(!proof.verify(leaf_hasher, path_hasher, &PH::Hash::one(), leaf));
+            assert!(!proof.verify(leaf_hasher, path_hasher, &PH::Hash::rand(&mut rng), leaf));
+        }
+    }
+
+    // If additional leaves are provided, check that the Merkle tree is consistent with them.
+    if !updates.is_empty() {
+        // Update the leaves of the Merkle tree.
+        for (leaf_index, leaf) in updates {
+            merkle_tree.update(*leaf_index, leaf)?;
+        }
+        // Check each additional leaf in the Merkle tree.
+        for (leaf_index, leaf) in updates {
+            // Compute a Merkle proof for the leaf.
+            let proof = merkle_tree.prove(*leaf_index, leaf)?;
+            // Verify the Merkle proof succeeds.
+            assert!(proof.verify(leaf_hasher, path_hasher, merkle_tree.root(), leaf));
+            // Verify the Merkle proof **fails** on an invalid root.
+            assert!(!proof.verify(leaf_hasher, path_hasher, &PH::Hash::zero(), leaf));
+            assert!(!proof.verify(leaf_hasher, path_hasher, &PH::Hash::one(), leaf));
+            assert!(!proof.verify(leaf_hasher, path_hasher, &PH::Hash::rand(&mut rng), leaf));
+        }
+    }
+    Ok(())
+}
+
+/// Runs the following test:
+/// 1. Construct a Merkle tree of a given depth with a given number of leaves.
+/// 2. Apply the updates to the Merkle tree.
+/// 3. Construct a new Merkle tree with the only the updated leaves.
+/// 4. Check that the Merkle root of the new Merkle tree is the same as the Merkle root of the original Merkle tree.
+fn check_updated_merkle_tree_is_consistent<
+    E: Environment,
+    LH: LeafHash<Hash = PH::Hash>,
+    PH: PathHash<Hash = Field<E>>,
+    const DEPTH: u8,
+>(
+    leaf_hasher: &LH,
+    path_hasher: &PH,
+    leaves: Vec<LH::Leaf>,
+    updates: Vec<(usize, LH::Leaf)>,
+) -> Result<()> {
+    // Construct the Merkle tree for the given leaves.
+    let mut merkle_tree = MerkleTree::<E, LH, PH, DEPTH>::new(leaf_hasher, path_hasher, &leaves)?;
+    assert_eq!(leaves.len(), merkle_tree.number_of_leaves);
+
+    // Construct an index map to track the leaves.
+    let mut leaf_map: IndexMap<usize, LH::Leaf> = leaves.into_iter().enumerate().collect::<IndexMap<usize, LH::Leaf>>();
+
+    // Apply the updates to the Merkle tree.
+    for (index, leaf) in &updates {
+        merkle_tree.update(*index, leaf)?;
+    }
+
+    // Add the updated leaves to the index map.
+    for (index, leaf) in updates {
+        leaf_map.insert(index, leaf);
+    }
+
+    // Get the updated leaves.
+    let updated_leaves = leaf_map.values().cloned().collect::<Vec<LH::Leaf>>();
```

Comment:
since it immediately gets collected into a vector, is the `IndexMap` for the `leaf_map` needed at all?

Reply:
The updates are done at random indices.
The `IndexMap` helps us keep track of them and get the unique updates.

Reply:
I meant as opposed to a plain `Vec` instead of an `IndexMap`

Reply:
Ah yes, we could use a `Vec` and dedup the keys.
I figured using the `IndexMap` would emulate the right sort of abstraction.